Fix mypy errors

20 errors exist in the code base as of the time of opening this issue.  

status_keeper.py:75: error: "str" has no attribute "name"
jockey.py:104: error: Item "None" of "Optional[Match[str]]" has no attribute "start"
jockey.py:108: error: Item "None" of "Optional[Match[str]]" has no attribute "group"
jockey.py:110: error: Item "None" of "Optional[Match[str]]" has no attribute "group"
jockey.py:112: error: Item "None" of "Optional[Match[str]]" has no attribute "end"
jockey.py:355: error: Return value expected
jockey.py:386: error: Missing return statement
jockey.py:409: error: Missing return statement
jockey.py:429: error: Argument 2 to "is_app_principal" has incompatible type "Optional[str]"; expected "str"
jockey.py:489: error: Argument 2 to "is_app_principal" has incompatible type "Optional[str]"; expected "str"
jockey.py:523: error: Missing return statement
jockey.py:568: error: Missing return statement
jockey.py:626: error: Argument 2 to "check_filter_match" has incompatible type "Optional[str]"; expected "str"
jockey.py:631: error: Argument 2 to "application_to_charm" has incompatible type "Optional[str]"; expected "str"
jockey.py:633: error: Argument 2 to "check_filter_match" has incompatible type "Optional[str]"; expected "str"
jockey.py:647: error: Argument 2 to "check_filter_match" has incompatible type "Optional[str]"; expected "str"
jockey.py:653: error: Argument 2 to "machine_to_hostname" has incompatible type "Optional[str]"; expected "str"
jockey.py:661: error: Argument 2 to "machine_to_ips" has incompatible type "Optional[str]"; expected "str"
jockey.py:694: error: Invalid index type "Optional[ObjectType]" for "Dict[ObjectType, Optional[Callable[[Dict[str, Any], List[JockeyFilter]], Generator[str, None, None]]]]"; expected type "ObjectType"
jockey.py:695: error: "None" not callable
Found 20 errors in 2 files (checked 2 source files)

--- a/jockey.py
+++ b/jockey.py
@@ -100,6 +100,7 @@ def parse_filter_string(
     assert len(filter_codes) == 1, "Incorrect number of filter codes detected."
 
     match = filter_code_pattern.search(filter_str)
+    assert match, "No filter code detected"
 
     object_type = convert_object_abbreviation(filter_str[: match.start()])
     assert object_type, "Invalid object type detected in filter string."
@@ -352,7 +353,7 @@ def application_to_charm(status: JujuStatus, app_name: str) -> Optional[str]:
     try:
         return status["applications"][app_name]["charm"]
     except KeyError:
-        return
+        return None
 
 
 def application_to_units(
@@ -405,6 +406,8 @@ def unit_to_application(status: JujuStatus, unit_name: str) -> Optional[str]:
     if app_name in status["applications"]:
         return app_name
 
+    return None
+
 
 def subordinate_unit_to_principal_unit(
     status: JujuStatus, unit_name: str
@@ -424,6 +427,7 @@ def subordinate_unit_to_principal_unit(
         The name of the unit to check.
     """
     app = unit_to_application(status, unit_name)
+    assert app, f"No application found for unit {unit_name}"
     app_data = status["applications"]
 
     if is_app_principal(status, app):
@@ -437,6 +441,8 @@ def subordinate_unit_to_principal_unit(
         for p_unit in app_data[p_app]["units"]:
             if unit_name in app_data[p_app]["units"][p_unit]["subordinates"]:
                 return p_unit
+
+    raise Exception(f"No principal unit detected for unit {unit_name}")
 
 
 def unit_to_machine(status: JujuStatus, unit_name: str) -> Optional[str]:
@@ -486,6 +492,7 @@ def machine_to_units(
 
         # Skip subordinate units
         app = unit_to_application(status, unit)
+        assert app
         if not is_app_principal(status, app):
             continue
 
@@ -540,6 +547,8 @@ def ip_to_machine(status: JujuStatus, ip: str) -> str:
         if ip in status["machines"][machine]["ip-addresses"]:
             return machine
 
+    raise Exception(f"No machine found with IP {ip}")
+
 
 def machine_to_hostname(status: JujuStatus, machine: str) -> str:
     """
@@ -585,6 +594,8 @@ def hostname_to_machine(status: JujuStatus, hostname: str) -> str:
         if status["machines"][machine]["hostname"] == hostname:
             return machine
 
+    raise Exception(f"No machine found for hostname {hostname}")
+
 
 def filter_units(
     status: JujuStatus, filters: List[JockeyFilter]
@@ -622,6 +633,7 @@ def filter_units(
         if app_filters or charm_filters:
             # Check application filters
             app = unit_to_application(status, unit)
+            assert app
             if not all(
                 check_filter_match(a_filter, app) for a_filter in app_filters
             ):
@@ -629,6 +641,7 @@ def filter_units(
 
             # Check charm filters
             charm = application_to_charm(status, app)
+            assert charm
             if not all(
                 check_filter_match(c_filter, charm)
                 for c_filter in charm_filters
@@ -643,6 +656,7 @@ def filter_units(
         # pdb.set_trace()
         # Check machine filters
         machine = unit_to_machine(status, unit)
+        assert machine
         if not all(
             check_filter_match(m_filter, machine)
             for m_filter in machine_filters
@@ -651,6 +665,7 @@ def filter_units(
 
         # Check hostname filters
         hostname = machine_to_hostname(status, machine)
+        assert hostname
         if not all(
             check_filter_match(h_filter, hostname)
             for h_filter in hostname_filters
@@ -659,6 +674,7 @@ def filter_units(
 
         # Check IP filters
         ips = machine_to_ips(status, machine)
+        assert ips
         if not all(
             any(check_filter_match(i_filter, ip) for ip in ips)
             for i_filter in ip_filters
@@ -690,8 +706,10 @@ def main(args: argparse.Namespace):
     }
 
     obj_type = convert_object_abbreviation(args.object)
+    assert obj_type, f"'{args.object}' is not a valid object type."
 
     action = RETRIEVAL_MAP[obj_type]
+    assert action
     print(" ".join(action(status, args.filters)))
 
 

--- a/status_keeper.py
+++ b/status_keeper.py
@@ -2,7 +2,8 @@ import datetime
 import json
 import os
 import subprocess
-from typing import Any, Dict
+from argparse import FileType
+from typing import Any, Dict, TextIO
 
 
 JOCKEY_PATH = os.path.expanduser("~/.jockey/")
@@ -68,11 +69,12 @@ def retrieve_juju_cache() -> Dict[str, Any]:
     return status
 
 
-def read_local_juju_status_file(filepath: str) -> Dict[str, Any]:
+def read_local_juju_status_file(file: TextIO) -> Dict[str, Any]:
     """
     Import Juju status from a local JSON file.
     """
-    filepath = os.path.abspath(filepath.name)
+    print("DEBUG:", type(file))
+    filepath = os.path.abspath(file.name)
     with open(filepath, "r") as file:
         status = json.loads(file.read())
 

